src/crypto/CHIPCryptoPALOpenSSL.cpp

```diff
@@ -1899,48 +1898,84 @@ CHIP_ERROR ExtractDNAttributeFromX509Cert(const char * oidString, const ByteSpan
 
-        if (strncmp(oidString, dnAttributeOidString, sizeof(dnAttributeOidString)) == 0)
+        bool isCommonNameObj = (OBJ_cmp(object, commonNameObj) == 0);
+        bool isMatterVidObj  = (OBJ_cmp(object, matterVidObj) == 0);
+        bool isMatterPidObj  = (OBJ_cmp(object, matterPidObj) == 0);
+
+        if (isCommonNameObj || isMatterVidObj || isMatterPidObj)
         {
             ASN1_STRING * data_entry = X509_NAME_ENTRY_get_data(name_entry);
             VerifyOrExit(data_entry != nullptr, err = CHIP_ERROR_INTERNAL);
             unsigned char * str = ASN1_STRING_data(data_entry);
             VerifyOrExit(str != nullptr, err = CHIP_ERROR_INTERNAL);
-
-            VerifyOrExit(ArgParser::ParseInt(reinterpret_cast<const char *>(str), id, 16), err = CHIP_ERROR_INTERNAL);
-            break;
+            int len = ASN1_STRING_length(data_entry);
+            VerifyOrExit(CanCastTo<size_t>(len), err = CHIP_ERROR_INTERNAL);
+
+            if (isMatterVidObj || isMatterPidObj)
+            {
+                uint16_t matterAttr;
+                VerifyOrExit(Encoding::UppercaseHexToUint16(reinterpret_cast<const char *>(str), static_cast<size_t>(len),
+                                                            matterAttr) == sizeof(matterAttr),
+                             err = CHIP_ERROR_WRONG_CERT_DN);
+
+                if (isMatterVidObj)
+                {
+                    // Not more than one VID attribute can be present.
+                    VerifyOrExit(!vidpid.mVendorId.HasValue(), err = CHIP_ERROR_WRONG_CERT_DN);
+                    vidpid.mVendorId.SetValue(static_cast<VendorId>(matterAttr));
+                }
+                else
+                {
+                    // Not more than one PID attribute can be present.
+                    VerifyOrExit(!vidpid.mProductId.HasValue(), err = CHIP_ERROR_WRONG_CERT_DN);
+                    vidpid.mProductId.SetValue(matterAttr);
+                }
+            }
+            // Otherwise, it is a CommonName attribute.
+            else if (!vidpidFromCN.mVendorId.HasValue() && !vidpidFromCN.mVendorId.HasValue())
+            {
+                char cnAttr[kMax_CommonNameAttr_Length + 1];
+                VerifyOrExit(static_cast<size_t>(len) <= chip::Crypto::kMax_CommonNameAttr_Length, err = CHIP_ERROR_INTERNAL);
+                memcpy(cnAttr, str, static_cast<size_t>(len));
+                cnAttr[len] = 0;
+
+                char * vid = strstr(cnAttr, kVIDPrefixForCNEncoding);
+                if (vid != nullptr)
+                {
+                    vid += strlen(kVIDPrefixForCNEncoding);
+                    VerifyOrExit(cnAttr + len >= vid + kVIDandPIDHexLength, err = CHIP_ERROR_INTERNAL);
+                    uint16_t matterAttr;
+                    VerifyOrExit(Encoding::UppercaseHexToUint16(vid, kVIDandPIDHexLength, matterAttr) == sizeof(matterAttr),
```

Comment:
Do we have a test case where the CN has a hex string that is not all uppercase, or wrong length, so that we don't transitively rely too too much on this without testing some of the outer assumption from Device Attestation part of spec?

We should have a `ExtractVidPidFromCn` utility that is *separately unit-tested* with all the examples from the spec.

Can be done as a follow-up

Reply:
Implemented a separate function to extract VID and PID from attribute string (could be CN or Matter specific RDN). Added tests to test all possible scenarios from the spec and some extra cases.